Add typing (PEP 561) marker file

ysangkok opened the pull request:
When type checking using mypy, I get errors when the package doesn't have a typing marker file.

Of course, this doesn't add actual type signatures, that would have to be added in addition. But since I don't know if you condone of PEP 561, I submit this PR. An alternative is to add the typing in https://github.com/python/typeshed , but it is better to have the typing info upstream. It won't affect Python 2 compatibility (typing can be in comments).

Errors without marker file:
```
electrum/dnssec.py:40: error: No library stub file for module 'dns'
electrum/dnssec.py:40: error: No library stub file for module 'dns.name'
electrum/dnssec.py:41: error: No library stub file for module 'dns.query'
electrum/dnssec.py:42: error: No library stub file for module 'dns.dnssec'
electrum/dnssec.py:43: error: No library stub file for module 'dns.message'
electrum/dnssec.py:44: error: No library stub file for module 'dns.resolver'
electrum/dnssec.py:45: error: No library stub file for module 'dns.rdatatype'
electrum/dnssec.py:46: error: No library stub file for module 'dns.rdtypes'
electrum/dnssec.py:46: error: No library stub file for module 'dns.rdtypes.ANY'
electrum/dnssec.py:46: error: No library stub file for module 'dns.rdtypes.ANY.NS'
electrum/dnssec.py:47: error: No library stub file for module 'dns.rdtypes.ANY.CNAME'
electrum/dnssec.py:48: error: No library stub file for module 'dns.rdtypes.ANY.DLV'
electrum/dnssec.py:49: error: No library stub file for module 'dns.rdtypes.ANY.DNSKEY'
electrum/dnssec.py:50: error: No library stub file for module 'dns.rdtypes.ANY.DS'
electrum/dnssec.py:51: error: No library stub file for module 'dns.rdtypes.ANY.NSEC'
electrum/dnssec.py:52: error: No library stub file for module 'dns.rdtypes.ANY.NSEC3'
electrum/dnssec.py:53: error: No library stub file for module 'dns.rdtypes.ANY.NSEC3PARAM'
electrum/dnssec.py:54: error: No library stub file for module 'dns.rdtypes.ANY.RRSIG'
electrum/dnssec.py:55: error: No library stub file for module 'dns.rdtypes.ANY.SOA'
electrum/dnssec.py:56: error: No library stub file for module 'dns.rdtypes.ANY.TXT'
electrum/dnssec.py:57: error: No library stub file for module 'dns.rdtypes.IN'
electrum/dnssec.py:57: error: No library stub file for module 'dns.rdtypes.IN.A'
electrum/dnssec.py:58: error: No library stub file for module 'dns.rdtypes.IN.AAAA'
```

pspacek commented:
Hello,

PEP 561 stub sounds like a very useful addition! We are using mypy ourselves and could use these typedefs.

rthalley commented:
I've got no issues with adding typeinfo at some point, though I haven't learned how to do that yet :)  Is it useful in the meantime to provide this stub file with no type content?  If so, glad to merge it.

ysangkok commented:
Because we'd need the `typing` module, real typeinfo could only be added if the minimum required Python version was 2.7.

pspacek commented:
Hmm, supported Python versions are recurring topic ... Looking at https://pythonclock.org/, Python 2 is going to be EOLed at the end of 2019, which is ~ year and half away.

Sooooo, is it worth maintaining anything for python <= 2.6?

Personally I do not think so and I've given up RHEL 6 a long time ago because it is impossible to compile new software on it anyway (too old GnuTLS etc.).

@rthalley what do you think? Can we move on to RHEL 7+ world, i.e. require Python 2.7?

ysangkok commented:
Sorry, I was wrong. As you can see, there are three ways to add annotations: https://www.python.org/dev/peps/pep-0561/#id17

The alternative that uses pyi would retain Python 2.6 compatibility. The type comment approach (not pyi) might even work if we catch the ImportError for 2.6. 

The third alternative is [PEP 484](https://www.python.org/dev/peps/pep-0484/) but that required Python 3, so we can rule that out.

The question is, should I work on the pyi approach or is it preferable with type comments? Testing the catching of ImportError, it seems to work.

pspacek commented:
@tomaskrizek What's your experience? Opinions?

rthalley commented:
Well, I was only keeping 2.6 support because RedHat requested it :)

Personally I'm happy enough to drop 2.7 and only support Python 3, but that's probably going too far!

/Bob

wagner-certat commented:
RHEL 7 also has python 3.4.

Python 3.3 is out of support too btw

ysangkok commented:
@rthalley I added some stubs, Mypy now runs without error on the codebase. A lot of work is still needed until it would pass with `--strict`. I removed the unittest2 dependency. Examples and tests use comment syntax, dns/ uses pyi files. Tell me what you think.

ysangkok commented:
@pspacek Are you running Mypy with `--strict`? This is how I invoke Mypy:

```
 % MYPYPATH=/usr/lib/python3/dist-packages/ mypy examples tests
tests/test_zone.py:476: error: Item "None" of "Optional[Rdataset]" has no attribute "ttl"
tests/test_zone.py:477: error: Value of type "Optional[Rdataset]" is not indexable
tests/test_zone.py:480: error: Item "None" of "Optional[Rdataset]" has no attribute "ttl"
tests/test_zone.py:483: error: Item "None" of "Optional[Rdataset]" has no attribute "ttl"
tests/test_zone.py:489: error: Item "None" of "Optional[Rdataset]" has no attribute "ttl"
tests/test_zone.py:492: error: Item "None" of "Optional[Rdataset]" has no attribute "ttl"
tests/test_zone.py:495: error: Item "None" of "Optional[Rdataset]" has no attribute "ttl"
```

The errors above are easily fixable by inserting calls to typing.cast which is simply a cast for the type system (no effect at runtime).

If I understand you correctly, you think that incomplete type annotations are problematic. So if I finished this, would you be in favour of merging?

rthalley commented:
I'd merge this but there are conflicts.

ysangkok commented:
@rthalley rebased. Please note that I added a typecheck in the Makefile/travis file, I removed the unittest2 dependency (it was for 2.6), and added the `typing` package (noop on Python 3.5+).
